### PR TITLE
Implement BatchNorm double backwards

### PR DESCRIPTION
**Pull request by gchanan**
The double backwards function is currently implemented as a python function called direction from C++, but the plan is to convert it to C++ code once ATen is integrated with autograd.

### CODE REVIEW

**Comment by gchanan** [comment]
The math isn't that clean here; I'm sure it can be simplified and made more efficient (as one example, I didn't use in-place functions).

File: torch/csrc/autograd/functions/batch_normalization.cpp [hunk 1]
```diff
@@ -186,12 +195,22 @@ auto BatchNormBackward::apply(const variable_list& grad_outputs) -> variable_lis
         eps);
   }
 
+  // Add saved variables used out of the pure autograd to inputs
+  variable_list all_inputs(grad_outputs);
+  all_inputs.push_back(input_var);
+  if (weight.get()) {
+    all_inputs.push_back(weight_var);
+    all_inputs.push_back(bias_var);
```
Comment (albanD) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: Biases are actually not used in this function.
Shouldn't they be removed from here and from the double backward function?

File: torch/nn/_functions/thnn/batchnorm_double_backwards.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+from torch.autograd.variable import Variable
+from functools import reduce
+from operator import mul
+
+
+def sum_exclude_dim1(to_sum, keepdim=True):
+    to_sum = to_sum.sum(dim=0, keepdim=True)
+    dim = 2
```
Comment (albanD) [inline torch/nn/_functions/thnn/batchnorm_double_backwards.py]: Is this line useful given the loop below?

File: torch/nn/_functions/thnn/batchnorm_double_backwards.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+        if ggB is not None:
+            ggB_expanded = expand_as_dim1(ggB, input)
+    else:
+        gamma_expanded = 1
+
+    # define some terms we will reuse
+    M = reduce(mul, input.size()[0:1] + input.size()[2:])
+    mu = sum_exclude_dim1(input).div_(M)
+    input_sub_mu = input - mu
```
Comment (albanD) [inline torch/nn/_functions/thnn/batchnorm_double_backwards.py]: Can this operation actually be made inplace in `mu`?

File: torch/nn/_functions/thnn/batchnorm_double_backwards.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    def first_back_grad_input(gO, gamma):
+        h0 = (gamma / (sigma2_eps).sqrt()).div_(M)
+        h1 = (M * gO).sub_(sum_exclude_dim1(gO)).sub_(input_sub_mu.div(sigma2_eps) * sum_exclude_dim1(gO * input_sub_mu))
+        return h0 * h1
+
+    # calculate gG
+    gG = None
+    if affine and ggI is not None:
+        # gG is just the first backwards with the gamma term removed (then shaped properly)
+        gG = ggI * first_back_grad_input(gO, 1)
```
Comment (albanD) [inline torch/nn/_functions/thnn/batchnorm_double_backwards.py]: Not sure how much you want to emphasize in the inplace part, but I think you could make this inplace as well as the return statement from `first_back_grad_input`.

File: torch/nn/_functions/thnn/batchnorm_double_backwards.py (hunk 3 above)
Comment (gchanan) [inline torch/nn/_functions/thnn/batchnorm_double_backwards.py]: It could, but from my reading of the API it would require two steps: i.e. mu.sub_(input).mul_(-1).

File: torch/nn/_functions/thnn/batchnorm_double_backwards.py (hunk 4 above)
Comment (gchanan) [inline torch/nn/_functions/thnn/batchnorm_double_backwards.py]: we don't currently support in-place multiplication of variables (because you'd have to save the old variable to compute the grad).

File: torch/nn/_functions/thnn/batchnorm_double_backwards.py (hunk 2 above)
Comment (gchanan) [inline torch/nn/_functions/thnn/batchnorm_double_backwards.py]: no, good catch.  This code also doesn't respect the keepdim parameter (and wouldn't use it correctly if it did).  Let me rewrite.

File: torch/csrc/autograd/functions/batch_normalization.cpp (hunk 1 above)
Comment (gchanan) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: yes, good point.  I was concerned that I wouldn't get a ggB but autograd handles that correctly :).

**Comment by gchanan** [comment]
updated for @albanD's suggestions.

**Comment by soumith** [comment]
woohoo! thanks Greg.
